sockeye/inference.py

```diff
@@ -1043,25 +1047,58 @@ def _beam_search(self,
             if self.restrict_lexicon:
                 best_word_indices[:] = vocab_slice_ids.take(best_word_indices)
 
-            # (4) get hypotheses and their properties for beam_size winning hypotheses (ascending)
+            # (4) Normalize the scores of newly finished hypotheses
+            newly_finished = (best_word_indices == self.vocab_target[C.EOS_SYMBOL]).expand_dims(axis=1)
+            scores_accumulated = mx.nd.where(newly_finished, scores_accumulated / self.length_penalty(lengths), scores_accumulated)
+
+            # (5) Prune out low-probability hypotheses
+            for sentno in range(self.batch_size):
+
+                # Set the active beam size to full. It will get reduced if items are filtered.
+                active_beam_size[sentno] = self.beam_size
+
+                rows = slice(sentno * self.beam_size, (sentno + 1) * self.beam_size)
+                if self.beam_prune > 0.0:
+                    if mx.nd.sum(finished[rows]) > 0:
+                        best_score = mx.nd.min(mx.nd.where(finished[rows].expand_dims(axis=1), scores_accumulated[rows], inf_array))
+
+                        to_remove = mx.nd.cast(scores_accumulated[rows,0] - best_score > self.beam_prune, dtype='int32')
+                        scores_accumulated[rows] = mx.nd.where(to_remove, inf_array, scores_accumulated[rows])
+                        best_word_indices[rows] = mx.nd.where(to_remove, zeros_array, best_word_indices[rows])
+
+                        num_removing = int(mx.nd.sum(to_remove).asscalar())
+                        active_beam_size[sentno] = self.beam_size - num_removing
+
+                        # mark removed ones as finished so they won't block early exiting
+                        finished[rows] = finished[rows] + to_remove
+
+                # Resort the hypotheses
+                sorted_indices = mx.nd.argsort(scores_accumulated[rows,0])
```

Comment:
I don't fully understand this statement. Why are you sorting based on the first column? If no pruning is applied (the block starting in line 1061 is not entered), do we still need this and why?

Reply:
Above the pruning, scores_accumulated applies pruning only to some (newly finished) hypotheses, so we have to re-sort.